### ISSUE TITLE
Compatibility of return type with subclass upon overload

### ISSUE THREAD

**Issue by cmp0xff** [python/mypy#19952]
Hi, we have the following minimal code snippet that reproduces our concern from pandas-dev/pandas-stubs#1385. In words, when the return types from different overloads are parametrised by a type variable, type checkers can work. When return types are subclasses, however, `mypy` does not seem to work.

```py
from typing import Any, Generic, Never, TypeVar, overload, reveal_type

T = TypeVar("T")

class Arg(Generic[T]): ...

class Gen(Generic[T]): ...

# part 1: return type is Generic

class D0:
    @overload
    def m(self, p: Arg[Never]) -> Gen: ...

    @overload
    def m(self, p: Arg[int]) -> Gen[int]: ...

reveal_type(D0().m(Arg[Any]()))  # pyright gives Gen[Unknown], mypy gives Gen[Any], good
reveal_type(D0().m(Arg[int]()))  # Gen[int], good
reveal_type(D0().m(Arg[float]()))  # error, good

# part 2: return types are subclasses

class Base: ...

class Sub(Base): ...

class D1:
    @overload
    def m(self, p: Arg[Never]) -> Base: ...  # we want to get Base for Arg[Any]; mypy warns overload-overlap here, whereas pyright does not warn (at least without strict)

    @overload
    def m(self, p: Arg[int]) -> Sub: ...

reveal_type(D1().m(Arg[Any]()))  # pyright gives Base, but mypy gives Any
reveal_type(D1().m(Arg[int]()))  # Sub, good
reveal_type(D1().m(Arg[float]()))  # error, good
```

Part 1 is understood from python/mypy#19525. Could someone explain part 2? Thank you 🙂

**Comment by ilevkivskyi** [comment]
When multiple overloads can be matched because of an `Any` argument(s), and the corresponding return types are not equivalent modulo `Any`, mypy intentionally infers `Any` as the result of the call. This is done to guarantee a fundamental rule of gradual typing: replacing a precise type with `Any` should not cause new errors. Consider this code:
```python
a: Arg[Any] = Arg[int]()
x: Sub = D1().m(a)
```
this code would generate an error, while it shouldn't.

**Comment by cmp0xff** [comment]
Hi @ilevkivskyi , thank you for triaging. 
> When multiple overloads can be matched because of an `Any` argument(s), and the corresponding return types are not equivalent modulo `Any`,

The complaint in the OP is that the return types should be considered as equivalent in the covariant sense, so that returning `Any` should not happen.
